artemis-journal/src/main/java/org/apache/activemq/artemis/core/journal/impl/JournalRecord.java

```diff
@@ -33,7 +31,15 @@
 
    private final int size;
 
-   private List<Pair<JournalFile, Integer>> updateFiles;
+   // use this singleton to save using a separated boolean field to mark the "deleted" state
```

Comment:
TBH I don't get the logic here..  

it seems that you are using some sort of auto-expandable int arrays instead of the previous List?


Can't we encapsulate the collections somehow it's a bit more understandable?


I won't be able to maintain this class any longer after this change. I want to get a clear udnerstanding and simplify the logic here.. even if we have to add an additional layer just to manage the auto-expanding collection here.

right now the data-structure is mixed with the logic here, and I don't understand what's going on.. perhaps it's just that I didn' get your logic and a video conference would help me understand it.

Reply:
> Can't we encapsulate the collections somehow it's a bit more understandable?

Good point, tomorrow I'll pack this in a more understandable version; in short, it contains 2 optimizations:
1) it accumulates any incoming addToUpdateFile while preserving the neg/pos count and size, if the last added file is unchanged
2) it amortize the footprint of a single entry ie 
Before: 24 bytes Pair + 16 bytes boxed Integer + 4 bytes array list entry
Now: 4 bytes JournalFile array slot + 8 bytes pos/neg count and size packed together

So, 44 bytes vs 12 bytes for each entry

Reply:
I've added a new commit to improve the readability of the proposed solution